proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,414 @@
+
+     type Sum2# :: TYPE r1 -> TYPE r2 -> TYPE (SumRep [r1, r2])
+     data Sum2# a b = (# a | #) | (# | b #)
+
+     type Sum3# :: TYPE r1 -> TYPE r2 -> TYPE r3 -> TYPE (SumRep [r1, r2, r3])
+     data Sum3# a b c = (# a | | #) | (# | b | #) | (# | | c #)
+
+     -- ...
+
+     type Sum64# :: TYPE r1 -> ... -> TYPE r64 -> TYPE (SumRep [r1, ..., r64])
+     data Sum64# ... = ...
+
+#. Change ``GHC.Types`` to have the following definition::
+
+     data List a = [] | a : List a
+
+#. Re-export ``List`` from ``GHC.List`` and ``GHC.Prelude``.
+
+#. Introduce a new extension ``-XListTuplePuns``; this extension is part
+   of ``-XHaskell98``, ``-XHaskell2010``, and ``-XGHC2021``. It is thus on by default.
+
+#. With ``-XListTuplePuns``:
+
+   1. An occurrence of ``[]`` in type-syntax (as defined in `#378`_) is a synonym
+      for ``GHC.List.List``.
+
+   #. An occurrence of ``[ty]`` in type-syntax is a synonym for ``GHC.List.List ty``.
+
+   #. An occurrence of ``()`` in type-syntax, where the type is not expected to be of kind ``Constraint``,
+      is a synonym for ``GHC.Tuple.Unit``.
+
+   #. An occurrence of ``(,,...,,)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
+      is a synonym for ``GHC.Tuple.Tuple``\ *n+1*.
+
+   #. An occurrence of ``(ty1,ty2,...,tyn-1,tyn)`` (for *n* ≧ 2) in type-syntax, where neither the type
+      is expected to be of kind ``Constraint`` and either none of the ``tyi`` are inferred to have kind ``Constraint``
+      or there exists a ``tyi`` inferred to kind ``Type`` and none of the ``tyj`` (with j < i) are inferred to have
+      kind ``Constraint``, is
+      a synonym for ``GHC.Tuple.Tuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.
+
+   #. With ``-XUnboxedTuples``, an occurrence of ``(# #)`` in type-syntax is a synonym for ``GHC.Exts.Unit#``.
+
+   #. With ``-XUnboxedTuples``, an occurrence of ``(#,,...,,#)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
+      is a synonym for ``GHC.Exts.Tuple``\ *n+1*\ ``#``.
+
+   #. With ``-XUnboxedTuples``, an occurrence of ``(# ty1, ty2, ... , tyn-1, tyn #)`` (for *n* ≧ 2) in type-syntax is a synonym
+      for ``GHC.Exts.Tuple``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.
+
+   #. With ``-XUnboxedSums``, an occurrence of ``(# | | ... | | #)`` where there are *n* pipes (for *n* ≧ 1) in type-syntax
+      is a synonym for ``GHC.Exts.Sum``\ *n+1*\ ``#``.
+
+   #. With ``-XUnboxedSums``, an occurrence of ``(# ty1 | ty2 | ... | tyn-1 | tyn #)`` (for *n* ≧ 2) in type-syntax is a
+      synonym for ``GHC.Exts.Sum``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.
+
+   #. An occurrence of ``()`` in type-syntax, where the type is expected to be of kind ``Constraint``,
+      is a synonym for ``GHC.Tuple.CUnit``.
+
+   #. An occurrence of ``(ty1, ty2, ..., tyn-1, tyn)`` (for *n* ≧ 2) in type-syntax, where the type is
+      expected to be of kind ``Constraint``, is a synonym for ``GHC.Tuple.CTuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.
```

Comment:
It is what we do today. It *is* scary. But my goal here is not to de-scarify this aspect of GHC -- just to change concrete syntax.

Reply:
I certainly don't intend for you to de-scarify anything :slightly_smiling_face: . Though maybe you want to remind the reader that GHC already does this (because tuples of `Constraint`-s and tuples of `Type`-s are already two different type constructors) and it's merely a syntax change.

Reply:
So noted.